### PR TITLE
Add StableDiffusionXLControlNetInpaintPipeline

### PR DESCRIPTION
**Pull request by viiika**
# What does this PR do?

Fixes # (issue)

Add a new class StableDiffusionXLControlNetInpaintPipeline() for using SDXL Inpaint Model with ControlNets.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_controlnet_inpaint.py [hunk 1]
```diff
@@ -0,0 +1,1547 @@
...
+                    latents = (1 - init_mask) * init_latents_proper + init_mask * latents
+
+                # call the callback, if provided
+                if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                    progress_bar.update()
+                    if callback is not None and i % callback_steps == 0:
+                        callback(i, t, latents)
+
+        # make sure the VAE is in float32 mode, as it overflows in float16
+        if self.vae.dtype == torch.float16 and self.vae.config.force_upcast:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_controlnet_inpaint.py:1452]: Can we update this part of the code to conform with the changes here: https://github.com/huggingface/diffusers/pull/4796 ?

**Review by patrickvonplaten — COMMENTED** [review]
Can we also add some tests?

File: src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_controlnet_inpaint.py (hunk 1 above)
Comment (viiika) [reply src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_controlnet_inpaint.py:1452]: Sure.

**Comment by viiika** [comment]
> Can we also add some tests?

Sure you can. I also left a test case in src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_controlnet_inpaint.py/#44 .

**Comment by viiika** [comment]
> is this same as #4694 ?

They are similar but not the same. This PR mainly adapted from src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl_inpaint.py, so it has the same coding style with pipeline_stable_diffusion_xl series. #4694 mainly adapted from controlnet, so it may have more similar class structure with controlnet instead of stable_diffusion_xl.

**Comment by deepconv** [comment]
Thanks for the work. I have tried to use the newly introduced class in this PR, but in my experiments, the inpainting result is often not compatible with the context. I'm using depth condition though.
For SD1.5, the inpainting model is a separate model trained in inpainting mode, instead of directly using the base SD1.5 model.
Can SDXL base model without special training work well for inpainting?

**Comment by yiyixuxu** [comment]
cc @harutatsuakiyama here

We only need one inpaiting pipeline for SDXL so it would be great if you guys could work together! We can make sure you are both the author of the same PR:) 

I think since we have already reviewed this PR https://github.com/huggingface/diffusers/pull/4694, maybe we will continue to work on this PR? @viiika  would you be willing to review it too and point out things that you implemented in your PR but are missing there?  

or, you guys decide to go with this PR that would be totally fine too!

Thanks

YiYi

**Comment by viiika** [comment]
> cc @harutatsuakiyama here
> 
> We only need one inpaiting pipeline for SDXL so it would be great if you guys could work together! We can make sure you are both the author of the same PR:)
> 
> I think since we have already reviewed this PR #4694, maybe we will continue to work on this PR? @viiika would you be willing to review it too and point out things that you implemented in your PR but are missing there?
> 
> or, you guys decide to go with this PR that would be totally fine too!
> 
> Thanks
> 
> YiYi

Besides, Refactor with a mask_image_processor https://github.com/huggingface/diffusers/pull/4444/files has also be updated. We may directly check this PR for faster merge.

**Pull request closed without merging**